libvast/vast/data.hpp

```diff
@@ -318,23 +313,28 @@ bool evaluate(const data& lhs, relational_operator op, const data& rhs);
 
 // -- convertible -------------------------------------------------------------
 
-bool convert(const list& xs, json& j);
-bool convert(const map& xs, json& j);
-bool convert(const record& xs, json& j);
-bool convert(const data& xs, json& j);
-
-/// Converts data with a type to "zipped" JSON, i.e., the JSON object for
-/// records contains the field names from the type corresponding to the given
-/// data.
-bool convert(const data& x, json& j, const type& t);
+template <class T, class... Opts>
+data to_data(const T& x, Opts&&... opts) {
+  data d;
+  if (convert(x, d, std::forward<Opts>(opts)...))
+    return d;
+  return {};
+}
 
 caf::error convert(const record& xs, caf::dictionary<caf::config_value>& ys);
 caf::error convert(const record& xs, caf::config_value& cv);
 caf::error convert(const data& x, caf::config_value& cv);
 
-caf::error convert(const caf::dictionary<caf::config_value>& xs, record& ys);
-caf::error convert(const caf::dictionary<caf::config_value>& xs, data& y);
-caf::error convert(const caf::config_value& x, data& y);
+bool convert(const caf::dictionary<caf::config_value>& xs, record& ys);
+bool convert(const caf::dictionary<caf::config_value>& xs, data& y);
+bool convert(const caf::config_value& x, data& y);
+
+// -- JSON -------------------------------------------------------------
+
+/// Prints data as JSON.
+/// @param x The data instance.
+/// @returns The JSON representation of *x*, or an error.
+caf::expected<std::string> to_json(const data& x);
```

Comment:
Thinking of these `to_string/to_json/to_yaml` family of function and having fmt already in vast I'd like to share a quick observation:

I think the whole engine of `printable` things can be rebuilt with fmt (thankfully vast printable machinery is very similar to the one in fmt), thus reducing maintenance costs in a long run (a wide community is interested to have FMT a bug-free and performant) and be more 3rd-party friendly (e.g. plugins authors would have a kind of "standard" approach to do printings with vast types). Also it would be more expressive and flexible to have possible rendering formats as options, something like the following:

```C++
auto d = vast::data{ make_something()};
auto text = fmt::format("{}", d);
auto json = fmt::format("{j}", d);
auto yaml = fmt::format("{y}", d);
```

Reply:
This is great observation. We actually had the very same discussion last week: what's the future of our concepts, in particular Printable, Parseable, Hashable. It's no question that FMT is a major win and probably our best for the implementation of what we call Printable right now.

A while ago, we wanted to have `to<yaml>`, `to<json>`, etc. for rendering a type in a specific way. (The `to` overloads return an `expected<T>` instead of a `T`.) But we are moving away from that. Now that `json` is gone, we're actually going in a different direction.

The underlying question is: what's the most flexible mechanism to print one type in different ways? By treating FMT a first-class library in the code base, your proposal makes a lot of sense. The only thing I wish we had is a symmetric API for parsing. I'm not sure if these are realistic expectations, though.

Reply:
We've had the discussion before @mavam, and I feel like forcing APIs to be symmetric will just hold us back.

For now, I think `to_json` is fine, but I agree with @ngrodzitski that a format option for {fmt} would be much more friendly to plugin authors.